File: app/src/androidTest/java/com/tadfisher/mvidemo/TestUtils.java
package com.tadfisher.mvidemo;

import static android.support.test.espresso.core.deps.guava.base.Preconditions.checkNotNull;

import android.support.design.widget.TextInputLayout;
import android.support.test.espresso.matcher.BoundedMatcher;
import android.view.View;
import org.hamcrest.Description;
import org.hamcrest.Matcher;

class TestUtils {

  static Matcher<View> hasTextInputLayoutErrorText(final Matcher<String> stringMatcher) {
    checkNotNull(stringMatcher);
    return new BoundedMatcher<View, TextInputLayout>(TextInputLayout.class) {

      @Override
      public void describeTo(Description description) {
        description.appendText("with error: ");
        stringMatcher.describeTo(description);
      }

      @Override
      public boolean matchesSafely(TextInputLayout textInputLayout) {
        CharSequence error = textInputLayout.getError();
        String errorString = error == null ? null : error.toString();
        return error != null && stringMatcher.matches(errorString);
      }
    };
  }
}

File: app/src/androidTest/kotlin/com/tadfisher/mvidemo/TestUtils.kt
package com.tadfisher.mvidemo

import android.support.design.widget.TextInputLayout
import android.support.test.espresso.matcher.BoundedMatcher
import android.view.View
import android.view.View.VISIBLE
import org.hamcrest.Description
import org.hamcrest.Matcher

fun hasTextInputLayoutErrorText(stringMatcher: Matcher<String>): Matcher<View> {
  checkNotNull(stringMatcher)
  return object : BoundedMatcher<View, TextInputLayout>(TextInputLayout::class.java) {

    override fun describeTo(description: Description) {
      description.appendText("with error: ")
      stringMatcher.describeTo(description)
    }

    public override fun matchesSafely(textInputLayout: TextInputLayout): Boolean {
      return stringMatcher.matches(
          if (textInputLayout.findViewById(R.id.textinput_error)?.visibility != VISIBLE)
            null else textInputLayout.error?.toString())
    }
  }
}
